## Loading the necessary libraries

In [ ]:
import pandas as pd
import numpy as np

## Loading the datasets

In [ ]:
df1 = pd.read_csv("/content/cells_geo.csv", delimiter = ';',)
df1.head()

Unnamed: 0   VILLES      STATUS  ... AREA     CELL_ID   SITE_CODE
0           0   ADJAME  In Service  ...  NaN  fbb2e75af3  0bb492ef17
1           1   ADJAME  In Service  ...  NaN  1502501d78  0bb492ef17
2           2   ADJAME  In Service  ...  NaN  0c2e182a0c  0bb492ef17
3           3  ASSINIE  In Service  ...  NaN  69bdfc69fd  8a1e6ce31f
4           4  ASSINIE  In Service  ...  NaN  7a1029d2ab  8a1e6ce31f

[5 rows x 12 columns]

In [ ]:
df2 = pd.read_csv("/content/Telcom_dataset.csv")
df2.head()

PRODUTC  VALUE               DATETIME  ... COUNTRY_B     CELL_ID     SITE_ID
0   Voice      0  2012-05-06 23:04:37.0  ...     19e2e  /TJNe+Mmtu  /TJNe+Mmtu
1     sms      0  2012-05-06 23:05:25.0  ...     19e2e  /+cKZKIp41  /+cKZKIp41
2     sms     25  2012-05-06 23:05:41.0  ...     19e2e  /u0FSD+ahi  /u0FSD+ahi
3   Voice     56  2012-05-06 23:05:42.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4   Voice     14  2012-05-06 23:06:29.0  ...     19e2e  051KhYDCpv  051KhYDCpv

[5 rows x 10 columns]

In [ ]:
df3 = pd.read_csv("/content/Telcom_dataset2.csv")
df3.head()

PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B     CELL_ID     SITE_ID
0     sms      0  2012-05-07 23:02:06.0  ...     19e2e  +854AcBQT2  +854AcBQT2
1   Voice      0  2012-05-07 23:03:44.0  ...     19e2e  +854AcBQT2  +854AcBQT2
2     sms      0  2012-05-07 23:04:06.0  ...     19e2e  +laSrk7g6q  +laSrk7g6q
3   Voice     37  2012-05-07 23:04:48.0  ...     19e2e  /xvRrCVKoQ  /xvRrCVKoQ
4     sms     25  2012-05-07 23:04:58.0  ...     19e2e  +bO+qkSonO  +bO+qkSonO

[5 rows x 10 columns]

In [ ]:
df4 = pd.read_csv("/content/Telcom_dataset3.csv")
df4.head()

PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B      CELLID     SIET_ID
0   Voice     61  2012-05-08 23:01:28.0  ...     19e2e  +38u2u/rfx  +38u2u/rfx
1     sms      0  2012-05-08 23:01:31.0  ...     19e2e  +Fs4mTvdKx  +Fs4mTvdKx
2   Voice     10  2012-05-08 23:02:59.0  ...     19e2e  +ZaaZozy/+  +ZaaZozy/+
3     sms      0  2012-05-08 23:03:04.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4   Voice     22  2012-05-08 23:03:16.0  ...     19e2e  /eeMklgyJA  /eeMklgyJA

[5 rows x 10 columns]

In [ ]:
df5 = pd.read_excel("/content/cells_geo_description.xlsx")
df5

Column name                                        Description  Format
0         VILLES                                               City  String
1         STATUS                                  In Service or not  String
2   LOCALISATION                                  in ABIDJAN or not  String
3     DECOUPZONE                                  Geographical Zone  String
4       ZONENAME                                       Name of Zone  String
5      LONGITUDE                                          Longitude   Float
6       LATITUDE                                           Latitude   Float
7         REGION                                             Region  String
8           AREA                                               Area  String
9        CELL_ID                                     ID of the cell  String
10     SITE_CODE  Site (there are several cells per site, severa...  String

In [ ]:
df6 = pd.read_excel("/content/CDR_description.xlsx")
df6

Column name                                        Description   Format
0          PRODUCT                                       Voice or SMS   String
1            VALUE                                      Billing price  Integer
2        DATE_TIME               Time in format yyyy-MM-dd hh:mm:ss.0   String
3     CELL_ON_SITE  Which cell in the site was used (not needed here)  Integer
4  DW_A_NUMBER_INT  Anonymized phone number of the person for whic...   String
5  DW_B_NUMBER_INT        Anonymized phone number of the counterparty   String
6        COUNTRY_A                  Country of party A (useless here)   String
7        COUNTRY_B                  Country of party B (useless here)   String
8          CELL_ID                                     ID of the cell   String
9          SITE_ID                                     ID of the SITE   String

## Merge the datasets

In [ ]:
#merge the datasets and 
#drop the missing values
data = [df1,df2,df3,df4]
merged_df = pd.concat(data, axis = 1)
df = merged_df.dropna()
df.head()

Unnamed: 0 VILLES      STATUS  ... COUNTRY_B      CELLID     SIET_ID
23        23.0  ABOBO  In Service  ...     19e2e  2q0YsSCxbH  2q0YsSCxbH
24        24.0  ABOBO  In Service  ...     19e2e  1MTLUFpdzd  1MTLUFpdzd
25        25.0  ABOBO  In Service  ...     19e2e  0/eYPFomxx  0/eYPFomxx
26        26.0  ABOBO  In Service  ...     19e2e  ++ctND+jGD  ++ctND+jGD
27        27.0  ABOBO  In Service  ...     OTHER  ++ctND+jGD  ++ctND+jGD

[5 rows x 42 columns]

## Data Exploration and Cleaning

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2604 entries, 23 to 3973
Data columns (total 42 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Unnamed: 0       2604 non-null   float64
 1   VILLES           2604 non-null   object 
 2   STATUS           2604 non-null   object 
 3   LOCALISATION     2604 non-null   object 
 4   DECOUPZONE       2604 non-null   object 
 5   ZONENAME         2604 non-null   object 
 6   LONGITUDE        2604 non-null   float64
 7   LATITUDE         2604 non-null   float64
 8   REGION           2604 non-null   object 
 9   AREA             2604 non-null   object 
 10  CELL_ID          2604 non-null   object 
 11  SITE_CODE        2604 non-null   object 
 12  PRODUTC          2604 non-null   object 
 13  VALUE            2604 non-null   int64  
 14  DATETIME         2604 non-null   object 
 15  CELL_ON_SITE     2604 non-null   object 
 16  DW_A_NUMBER_INT  2604 non-null   object 
 17  DW_B_NUMBER_I

### Drop unnecessary columns

In [ ]:
to_drop = ['Unnamed: 0' ,'SIET_ID', 'CELLID', 'DW_A_NUMBER_INT', 'DW_B_NUMBER_INT', 'LATITUDE', 'LONGITUDE', 'CELL_ID', 'COUNTRY_A', 'COUNTRY_B', 'DW_A_NUMBER', 'DW_B_NUMBER']
df.drop(to_drop, axis = 1, inplace=True)
df.head()

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4913: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


VILLES      STATUS LOCALISATION  ... VALUE              DATE_TIME CELL_ON_SITE
23  ABOBO  In Service      ABIDJAN  ...    25  2012-05-08 23:09:48.0            2
24  ABOBO  In Service      ABIDJAN  ...     0  2012-05-08 23:10:16.0            4
25  ABOBO  In Service      ABIDJAN  ...     0  2012-05-08 23:10:18.0            0
26  ABOBO  In Service      ABIDJAN  ...    25  2012-05-08 23:10:19.0            0
27  ABOBO  In Service      ABIDJAN  ...     1  2012-05-08 23:10:20.0            _

[5 rows x 22 columns]

### Check for missing values

In [ ]:
# Check for null values if merged dataframe
df.isnull().sum()

VILLES          0
STATUS          0
LOCALISATION    0
DECOUPZONE      0
ZONENAME        0
REGION          0
AREA            0
SITE_CODE       0
PRODUTC         0
VALUE           0
DATETIME        0
CELL_ON_SITE    0
SITE_ID         0
PRODUCT         0
VALUE           0
DATE_TIME       0
CELL_ON_SITE    0
SITE_ID         0
PRODUCT         0
VALUE           0
DATE_TIME       0
CELL_ON_SITE    0
dtype: int64

In [ ]:
#RENAME COLUMN PRODUTC
df.rename({'PRODUTC': 'PRODUCT'}, axis=1)

VILLES      STATUS LOCALISATION  ... CELL_ON_SITE        date      time
23       ABOBO  In Service      ABIDJAN  ...            2  2012-05-06  23:11:27
24       ABOBO  In Service      ABIDJAN  ...            4  2012-05-06  23:11:27
25       ABOBO  In Service      ABIDJAN  ...            0  2012-05-06  23:12:07
26       ABOBO  In Service      ABIDJAN  ...            0  2012-05-06  23:12:10
27       ABOBO  In Service      ABIDJAN  ...            _  2012-05-06  23:12:25
...        ...         ...          ...  ...          ...         ...       ...
3968  ZUENOULA  In Service    INTERIEUR  ...            1  2012-05-07  00:01:26
3969  ZUENOULA  In Service    INTERIEUR  ...            3  2012-05-07  00:01:26
3970  ZUENOULA  In Service    INTERIEUR  ...            1  2012-05-07  00:01:26
3972  ZUENOULA  In Service    INTERIEUR  ...            0  2012-05-07  00:01:26
3973  ZUENOULA  In Service    INTERIEUR  ...            2  2012-05-07  00:01:26

[2604 rows x 24 columns]

In [ ]:
#REMOVING DUPLICATE COLUMNS TO REMAIN WITH ONE
df.loc[:, ~df.columns.duplicated()]

VILLES      STATUS  ...        date      time
23       ABOBO  In Service  ...  2012-05-06  23:11:27
24       ABOBO  In Service  ...  2012-05-06  23:11:27
25       ABOBO  In Service  ...  2012-05-06  23:12:07
26       ABOBO  In Service  ...  2012-05-06  23:12:10
27       ABOBO  In Service  ...  2012-05-06  23:12:25
...        ...         ...  ...         ...       ...
3968  ZUENOULA  In Service  ...  2012-05-07  00:01:26
3969  ZUENOULA  In Service  ...  2012-05-07  00:01:26
3970  ZUENOULA  In Service  ...  2012-05-07  00:01:26
3972  ZUENOULA  In Service  ...  2012-05-07  00:01:26
3973  ZUENOULA  In Service  ...  2012-05-07  00:01:26

[2604 rows x 17 columns]

### Split time_date to time and date 

In [ ]:
#date
df['date'] = pd.to_datetime(df['DATETIME']).dt.date

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [ ]:
#time
df['time'] = pd.to_datetime(df['DATETIME']).dt.time

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [ ]:
df.columns

Index(['VILLES', 'STATUS', 'LOCALISATION', 'DECOUPZONE', 'ZONENAME', 'REGION',
       'AREA', 'SITE_CODE', 'PRODUTC', 'VALUE', 'DATETIME', 'CELL_ON_SITE',
       'SITE_ID', 'PRODUCT', 'VALUE', 'DATE_TIME', 'CELL_ON_SITE', 'SITE_ID',
       'PRODUCT', 'VALUE', 'DATE_TIME', 'CELL_ON_SITE', 'date', 'time'],
      dtype='object')

In [ ]:
df.sample(5)

VILLES      STATUS LOCALISATION  ... CELL_ON_SITE        date      time
2044         KONG  In Service    INTERIEUR  ...            1  2012-05-07  00:00:33
2307       MAFERE  In Service    INTERIEUR  ...            1  2012-05-07  00:00:40
3475   TIEMELEKRO  In Service    INTERIEUR  ...            2  2012-05-07  00:01:13
1035  BROFFOUEDOU  In Service    INTERIEUR  ...            4  2012-05-07  00:00:08
1233        DALOA  In Service    INTERIEUR  ...            2  2012-05-07  00:00:12

[5 rows x 24 columns]

## Getting insights from our merged dataset

In [ ]:
#maximum billing price
df['VALUE'].max()

VALUE    1610
VALUE    3380
VALUE    6750
dtype: int64

In [ ]:
#minimum billing price
df['VALUE'].min()

VALUE    0
VALUE    0
VALUE    0
dtype: int64

In [ ]:
#mean billing price
df['VALUE'].mean()

VALUE    27.522657
VALUE    26.953917
VALUE    26.334869
dtype: float64

**1. Which ones were the most used city for the three days?**

In [ ]:
# Getting the count of the column villes('cities')
#
result = pd.value_counts(df['VILLES'])
city_users = result.head()
c_max = result.max()
print(f"The city with the most users appears: {c_max} times\n")
print(f"The Most used cities in 3 days: \n CITY       COUNT\n{city_users}")

The city with the most users appears: 241 times

The Most used cities in 3 days: 
 CITY       COUNT
COCODY      241
ABOBO       211
YOPOUGON    184
MARCORY      96
ADJAME       81
Name: VILLES, dtype: int64


**2. Which region is most used for the 3 days**

In [ ]:
# Getting the region that has most users.
#
region = pd.value_counts(df['REGION'])
print(f"\nThe most used region in 3 days are:\n")
print(f"Region               Count\n\
-------------------------\n{region.head()}")


The most used region in 3 days are:

Region               Count
-------------------------
LAGUNE               1290
BAS SASSANDRA         155
VALLEE DU BANDAMA     101
LACS                   96
SAVANE                 95
Name: REGION, dtype: int64


**3. Number of users in Abidjan and those outside of Abidjan.**

In [ ]:
# Getting the list of cities in a particular region.
c_loc = pd.value_counts(df['LOCALISATION'])
print(c_loc)

INTERIEUR    1431
ABIDJAN      1173
Name: LOCALISATION, dtype: int64


**4. Number of Users using Text, Voice and Data in COCODY City.**

In [ ]:
# Getting the Services data for MTN users in COCODY city.
# Voice
# data
# sms
#
voice = df[(df['VILLES'] == 'ABOBO') & (df['PRODUCT'] == "Voice")]
data = df[(df['VILLES'] == 'ABOBO') & (df['PRODUCT'] == "data")]
sms = df[(df['VILLES'] == 'ABOBO') & (df['PRODUCT'] == "sms")]

voice_count = pd.value_counts(voice['PRODUCT'])
data_count = pd.value_counts(data['PRODUCT'])
sms_count = pd.value_counts(sms['PRODUCT'])

print("The Number of Users who Accessed MTN Services in \
\n       ABOBO city\n-----------------------------\n\
        In 3 Days\n-----------------------------")
print(f"Total Voice users: {voice_count}\n\nTotal data users: {data_count}\n\nTotal sms users: {sms_count}\n\n")

**6. Sum of Billing Price used for the 3 days.**



Billing Price In Cococdy City.

In [ ]:
v_cocody = df[df['VILLES'] == 'COCODY']
sum_value = v_cocody['VALUE'].sum()
print(f"The sum of the Billing price per day in Cocody city is\n\
------------------------------------------------------\n{sum_value}")

The sum of the Billing price per day in Cocody city is
------------------------------------------------------
VALUE     5990
VALUE    11296
VALUE    13045
dtype: int64


Billing Price In Abobo City.

In [ ]:
v_abobo = df[df['VILLES'] == 'ABOBO']
sum_value = v_abobo['VALUE'].sum()
print(f"The sum of the Billing price per day in Abobo city is\n\
------------------------------------------------------\n{sum_value}")

The sum of the Billing price per day in Abobo city is
------------------------------------------------------
VALUE    3965
VALUE    3612
VALUE    3489
dtype: int64


#### Observations


Cocody City had the most Number of users. 



Abobo City had the second most Number of users.



Most Used regions. LAGUNE with 1290, BAS SASSANDRA with 155, VALLEE DU BANDAMA with 101.



Most users were from outside Abidjan Region with Abidjan having : 1173 and INTERIEUR : 1431

